scheduler/scheduler-examples/src/main/java/org/ow2/proactive/scheduler/examples/SFTPConnector.java

```diff
@@ -0,0 +1,396 @@
+
+import org.ow2.proactive.scheduler.common.task.TaskResult;
+import org.ow2.proactive.scheduler.common.task.executable.JavaExecutable;
+import org.zeroturnaround.zip.ZipUtil;
+
+import com.jcraft.jsch.Channel;
+import com.jcraft.jsch.ChannelSftp;
+import com.jcraft.jsch.JSch;
+import com.jcraft.jsch.JSchException;
+import com.jcraft.jsch.Session;
+import com.jcraft.jsch.SftpException;
+
+
+/**
+ * Import and export file(s) using SFTPConnector server.
+ *
+ * @author The ProActive Team
+ */
+public class SFTPConnector extends JavaExecutable {
+
+    private static final String DEFAULT_SFTP_HOSTNAME = "localhost";
+
+    private static final int DEFAULT_SFTP_PORT = 22;
+
+    private String sftpHostname = DEFAULT_SFTP_HOSTNAME;
+
+    private int sftpPort = DEFAULT_SFTP_PORT;
+
+    private static final String GET = "GET";
+
+    private static final String PUT = "PUT";
+
+    private static final String STRICT_HOST_KEY_CHEKING = "StrictHostKeyChecking";
+
+    private static final String CURRENT_FOLDER = ".";
+
+    private static final String PARENT_FOLDER = "..";
+
+    private boolean sftpExtractArchive;
+
+    private String sftpLocalRelativePath;
+
+    private String sftpRemoteRelativePath;
+
+    private String sftpMode;
+
+    private String sftpUsername;
+
+    private String sftpUrlKey;
+
+    private String sftpPassword;
+
+    private static final String SFTP_LOCAL_RELATIVE_PATH_ARG = "sftpLocalRelativePath";
+
+    @Override
+    public void init(Map<String, Serializable> args) throws Exception {
+        if (args.containsKey("sftpHostname")) {
+            sftpHostname = args.get("sftpHostname").toString();
+        }
+        if (args.containsKey("sftpPort")) {
+            try {
+                sftpPort = Integer.parseInt(args.get("sftpPort").toString());
+            } catch (NumberFormatException e) {
+                getOut().println("Invalid port number. Using default sftp port=" + DEFAULT_SFTP_PORT);
+            }
+        }
+        if (args.containsKey("sftpMode")) {
+            sftpMode = args.get("sftpMode").toString();
+        }
+        if (args.containsKey("sftpExtractArchive")) {
+            sftpExtractArchive = Boolean.parseBoolean(args.get("sftpExtractArchive").toString());
+        }
+        if (args.containsKey(SFTP_LOCAL_RELATIVE_PATH_ARG) &&
+            !args.get(SFTP_LOCAL_RELATIVE_PATH_ARG).toString().isEmpty()) {
+            sftpLocalRelativePath = args.get(SFTP_LOCAL_RELATIVE_PATH_ARG).toString();
+        } else {
+            //we throw an exception tp prevent transferring all the contents of the global space.
+            if (sftpMode.equals(PUT)) {
+                throw new IllegalArgumentException("Please specify a local relative path. Empty value is not allowed.");
+            }
+            //default value is getlocalspace() because it will always be writable and moreover can be used to transfer files to another data space(global, user)
+            sftpLocalRelativePath = getLocalSpace();
```

Comment:
getLocalSpace() is an absolute path.

instead of using this as the default sftpLocalRelativePath, you should define a sftpLocalAbsolutePath:

- if sftpLocalRelativePath is empty, then sftpLocalAbsolutePath = getLocalSpace()
- if sftpLocalRelativePath is not empty, then sftpLocalAbsolutePath = new File(getLocalSpace(), sftpLocalRelativePath)

This is the only way to guaranty coherence between fork or non-fork mode. In the latter case, current directory (.) is not writeable, but getLocalSpace() is always writeable.

Reply:
Makes sense. Done